src/ucp/core/ucp_worker.c

```diff
@@ -2339,6 +2339,17 @@ static void ucp_worker_destroy_eps(ucp_worker_h worker)
     ucs_debug("worker %p: destroy all endpoints", worker);
     ucs_list_for_each_safe(ep_ext, tmp, &worker->all_eps, ep_list) {
         ep = ucp_ep_from_ext_gen(ep_ext);
+        if (ucp_ep_config(ep)->key.err_mode == UCP_ERR_HANDLING_MODE_PEER) {
```

Comment:
what if error handling mode is NONE, but pendings are present?

Reply:
it will be failed with fatal error in the callback - and printing a message in `ucs_bug()`

Reply:
I think we don't need to take care of such EPs. Normally, user's EPs should not go here.

Reply:
so user is not allowed to destroy worker until all eps are flushed/closed?

Reply:
that is not allowed for any err_mode

Reply:
yes, but anyway if error handling was requested, we can purge pending/outstanding operations to a user

Reply:
@dmitrygx didn't we say it's no longer needed after UD fixes?
what happens if we remove the check?

Reply:
> @dmitrygx didn't we say it's no longer needed after UD fixes?

unfortunately, I found that UD moves purging outstanding operations from UD iface to EP, but if the operations are pending, so these requests will be purged with a callback that calls `ucs_bug()`

> what happens if we remove the check?

crash with XPMEM (it touches remote peer memory which couldn't be available anymore) which doesn't support error handling and hence flush(CANCEL)

Reply:
should not workers to be flushed before destroy if err handling was not requested?

Reply:
updated the PR to implement and use `ucp_ep_purge_lanes()` instead of `ucp_ep_discard_lanes()` which does unnecessary action - flush(CANCEL).
btw, there are two places where we do purge of pending request in a loop for all EP lanes, but also it does some other actions in the same loop. Shall we split this loop into two loops: 1) pending purging, 2) other actions, e.g. destroying UCT EP? The 1) action will done by calling `ucp_ep_purge_lanes()`. wdyt?

Reply:
> should not workers to be flushed before destroy if err handling was not requested?

Yes, we should do it. But we don't wait for Reply EP being completed its work, e.g. by incrementing/decrementing flush_ops counter in UCP Worker

Reply:
why? IMO all hidden eps (reply, KA, etc) should part of worker flush as well

Reply:
> why? IMO all hidden eps (reply, KA, etc) should part of worker flush as well

yes, if we do flush all EPs as well

Reply:
we don't require flushing the worker before destroying it

Reply:
but then we have eps  (KA) which is not handled as all others, i. e. not flushed (which requires special handling, like bug fix here).

Reply:
the EP created for sending EP_REMOVED answer during KA is created only if error handling requested by user.
E.g. We don't do flush in gtests for error handling tests

Reply:
we can silently destroy such KA-reply temporary ep and cancel all its operations internally and not wait for them to complete since we are going to destroy all transports anyway

Reply:
AFAIU, KA-reply EPs has to be created only on UCT ifaces which support peer failure err handling (we should request it on internal ep_create(ka_ep)), so we can assert or I'm missing something?

Reply:
I've removed the check here
So, now we purge pending requests for all EPs (KA-reply EPs, EPs which weren't exposed to a user, EPs which weren't destroyed y a user).